tests/app/notifications/test_process_notification.py

```diff
@@ -0,0 +1,90 @@
+
+
+def test_persist_notification_creates_and_save_to_db(sample_template, sample_api_key):
+    assert Notification.query.count() == 0
+    assert NotificationHistory.query.count() == 0
+    notification = persist_notification(sample_template.id, sample_template.version, '+447111111111',
+                                        sample_template.service.id, {}, 'sms', sample_api_key.id,
+                                        sample_api_key.key_type)
+    assert Notification.query.count() == 1
+    assert Notification.query.get(notification.id).__eq__(notification)
+    assert NotificationHistory.query.count() == 1
+
+
+def test_persist_notification_throws_exception_when_missing_template(sample_template, sample_api_key):
+    assert Notification.query.count() == 0
+    assert NotificationHistory.query.count() == 0
+    with pytest.raises(SQLAlchemyError):
+        persist_notification(template_id=None,
+                             template_version=None,
+                             recipient='+447111111111',
+                             service_id=sample_template.service.id,
+                             personalisation=None, notification_type='sms',
+                             api_key_id=sample_api_key.id,
+                             key_type=sample_api_key.key_type)
+    assert Notification.query.count() == 0
+    assert NotificationHistory.query.count() == 0
+
+
+@pytest.mark.parametrize('research_mode, queue, notification_type, key_type',
+                         [(True, 'research-mode', 'sms', 'normal'),
+                          (True, 'research-mode', 'email', 'normal'),
+                          (True, 'research-mode', 'email', 'team'),
+                          (False, 'send-sms', 'sms', 'normal'),
+                          (False, 'send-email', 'email', 'normal'),
+                          (False, 'send-sms', 'sms', 'team'),
+                          (False, 'research-mode', 'sms', 'test')])
+def test_send_notification_to_queue(notify_db, notify_db_session,
+                                    research_mode, notification_type,
+                                    queue, key_type, mocker):
+    mocked = mocker.patch('app.celery.provider_tasks.deliver_{}.apply_async'.format(notification_type))
+    template = sample_template(notify_db, notify_db_session) if notification_type == 'sms' \
+        else sample_email_template(notify_db, notify_db_session)
+    notification = sample_notification(notify_db, notify_db_session, template=template, key_type=key_type)
+    send_notification_to_queue(notification=notification, research_mode=research_mode)
+
+    mocked.assert_called_once_with([str(notification.id)], queue=queue)
+
+
+def test_send_notification_to_queue_throws_exception_deletes_notification(sample_notification, mocker):
+    mocked = mocker.patch('app.celery.provider_tasks.deliver_sms.apply_async', side_effect=Boto3Error("EXPECTED"))
+    with pytest.raises(Boto3Error):
```

Comment:
The code catches Exception, so does it matter what exception I set the mock to throw?